### PR TITLE
Update fetch methods

### PR DESCRIPTION
**Pull request by KiChjang**
Includes updates to main fetch, scheme fetch, HTTP fetch, HTTP-redirect fetch, HTTP-network fetch, HTTP-network-or-cache fetch and CORS preflight fetch.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17521)

### CODE REVIEW

File: components/net/fetch/methods.rs [hunk 1]
```diff
@@ -177,35 +176,46 @@ pub fn main_fetch(request: &mut Request,
 
         if (same_origin && !cors_flag ) ||
             current_url.scheme() == "data" ||
-            current_url.scheme() == "file" ||
```
Comment (jdm) [inline components/net/fetch/methods.rs]: I am 99% sure this is going to hit the same problem as https://github.com/servo/servo/pull/14515#issuecomment-266173937 .

Comment (KiChjang) [inline components/net/fetch/methods.rs]: How unfortunate. Looks like I'll have to drop those changes then.

**Comment by KiChjang** [comment]
Huh... looks like this update failed some unit tests, will look into them.

**Comment by KiChjang** [comment]
I found out what's wrong with the tests, and it's because of the redirection code that lives in `network_listener.rs`. Since `http_redirect_fetch` does not check the headers to see whether there is a `Location` header anymore, we will need to create a `location_url` field on `ResponseInit`. I'm currently stuck on that, because `location_url` needs to implement `serde::Serialize` and `serde::Deserialize`, which means `url::ParseError` needs to have those implementations as well, but it is not quite simple implementing those two traits on `url::ParseError`.

**Comment by KiChjang** [comment]
Ok, instead of storing an `Option<Result<ServoUrl, url::ParseError>>`, I'm replacing the `ParseError` with a `String`.

@bors-servo try

File: components/net/fetch/methods.rs [hunk 2]
```diff
@@ -216,9 +229,28 @@ pub fn main_fetch(request: &mut Request,
...
-    let mut response = if response.response_type == ResponseType::Default {
+    let mut response = if !response.is_network_error() && response.internal_response.is_none() {
+        // Substep 1.
+        if request.response_tainting == ResponseTainting::CorsTainting {
+            // Subsubstep 1.
+            let header_names = response.headers.get::<AccessControlExposeHeaders>();
+            match header_names {
+                // Subsubstep 2.
+                Some(list) if request.credentials_mode != CredentialsMode::Include => {
+                    if list.iter().any(|s| **s == "*".to_owned()) {
```
Comment (jdm) [inline components/net/fetch/methods.rs]: We should be checking if list.len() == 1 and the first element is "*".

File: components/net/fetch/methods.rs [hunk 3]
```diff
@@ -476,7 +508,8 @@ pub fn is_simple_header(h: &HeaderView) -> bool {
     }
 }
 
-pub fn is_simple_method(m: &Method) -> bool {
+// https://fetch.spec.whatwg.org/#cors-safelisted-method
```
Comment (jdm) [inline components/net/fetch/methods.rs]: ///

File: components/net/http_loader.rs [hunk 4]
```diff
@@ -682,7 +673,8 @@ pub fn http_redirect_fetch(request: &mut Request,
     request.redirect_count += 1;
 
     // Step 7
-    let same_origin = location_url.origin()== request.current_url().origin();
+    // FIXME: Correctly use request's origin
+    let same_origin = location_url.origin() == request.current_url().origin();
```
Comment (jdm) [inline components/net/http_loader.rs]: Let's use the correct origin here. We can [create](https://doc.servo.org/servo_url/origin/enum.ImmutableOrigin.html#method.new) an ImmutableOrigin value from the url's origin and compare them.

File: components/net/http_loader.rs [hunk 5]
```diff
@@ -695,28 +687,38 @@ pub fn http_redirect_fetch(request: &mut Request,
...
 
-    // Step 10
-    let status_code = response.actual_response().status.unwrap();
-    if ((status_code == StatusCode::MovedPermanently || status_code == StatusCode::Found) &&
-        request.method == Method::Post) ||
-        status_code == StatusCode::SeeOther {
+    // Step 11
+    if response.actual_response().status.map_or(false, |code|
+        ((code == StatusCode::MovedPermanently || code == StatusCode::Found) && request.method == Method::Post) ||
+        code == StatusCode::SeeOther) {
```
Comment (jdm) [inline components/net/http_loader.rs]: I don't think that this change is an improvement. I found it easier to read the code before.

File: components/net/http_loader.rs [hunk 6]
```diff
@@ -935,82 +937,73 @@ fn http_network_or_cache_fetch(request: &mut Request,
...
-            // Step 4
-            // return http_network_or_cache_fetch(request, authentication_fetch_flag,
-            //                                    cors_flag, done_chan, context);
-        },
-        _ => {}
+        // Step 3
+        // TODO: Prompt the user for proxy authentication credentials
+        // Wrong, but will have to do until we are able to prompt the user
+        // otherwise this creates an infinite loop
+        // We basically pretend that the user declined to enter credentials
```
Comment (jdm) [inline components/net/http_loader.rs]: This comment is incorrect if we don't return response like the old code did.

File: components/net/http_loader.rs [hunk 7]
```diff
@@ -1202,94 +1203,115 @@ fn cors_preflight_fetch(request: &Request,
     preflight.initiator = request.initiator.clone();
     preflight.type_ = request.type_.clone();
     preflight.destination = request.destination.clone();
+    preflight.origin = request.origin.clone();
```
Comment (jdm) [inline components/net/http_loader.rs]: Good catch!

File: components/net/http_loader.rs [hunk 8]
```diff
@@ -1202,94 +1203,115 @@ fn cors_preflight_fetch(request: &Request,
...
                 None => return Response::network_error(NetworkError::Internal("CORS ACAH check failed".into()))
             }
         } else {
             vec![]
         };
 
-        // Substep 4
+        // Substep 5
+        if (methods.iter().any(|m| m.as_ref() == "*") ||
+            header_names.iter().any(|hn| **hn == "*".to_owned())) &&
```
Comment (jdm) [inline components/net/http_loader.rs]: Can we do `***hn == "*"` instead?

File: components/net/http_loader.rs [hunk 9]
```diff
@@ -1202,94 +1203,115 @@ fn cors_preflight_fetch(request: &Request,
...
-            !is_cors_safelisted_method(&request.method) {
+            !is_cors_safelisted_method(&request.method) &&
+            methods.iter().all(|m| m.as_ref() != "*") {
             return Response::network_error(NetworkError::Internal("CORS method check failed".into()));
         }
 
-        // Substep 6
+        // Substep 8
+        if request.headers.iter().any(
+            |header| header.name().to_ascii_lowercase() == "authorization" &&
```
Comment (jdm) [inline components/net/http_loader.rs]: No need for lowercasing; UniCase ignores string case when comparing.

File: components/net/http_loader.rs [hunk 10]
```diff
@@ -1202,94 +1203,115 @@ fn cors_preflight_fetch(request: &Request,
...
+                     header_names.iter().all(|hn| *hn != UniCase(header.name().to_owned()))) {
```
Comment (jdm) [inline components/net/http_loader.rs]: Can we do UniCase(header.name()) instead?

**Review by jdm — CHANGES_REQUESTED** [review]
Thanks for doing this work!

File: components/net/http_loader.rs (hunk 4 above)
Comment (KiChjang) [inline components/net/http_loader.rs]: That shouldn't be neccessary either. All code paths that come from `RequestInit` should already have an `ImmutableOrigin` set for them.

File: components/net/http_loader.rs (hunk 5 above)
Comment (KiChjang) [inline components/net/http_loader.rs]: The improvement here is that the code doesn't panic when it encounters a None on the status field of the actual response.

**Comment by KiChjang** [comment]
Force-pushed with comments addressed.

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
```
  ▶ TIMEOUT [expected PASS] /_mozilla/css/transition_calc_implicit.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ Unexpected subtest result in /cors/redirect-userinfo.htm:
  │ FAIL [expected PASS] Allow redirect without userinfo (@ is trimmed during URL parsing)
  └   → assert_unreached: Reached unreachable code

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  │ FAIL [expected PASS] remote (*) to local (http://web-platform.test:8000), expect to fail
  │   → assert_false: Got response expected false got true
  │ FAIL [expected PASS] remote (http://web-platform.test:8000) to local (http://web-platform.test:8000), expect to fail
  │   → assert_false: Got response expected false got true
  │ 
  │ shouldFail/</client.onreadystatechange<@http://web-platform.test:8000/cors/redirect-origin.htm:162:17
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  │ FAIL [expected PASS] remote (*) to remote (http://web-platform.test:8000), expect origin=http://web-platform.test:8000
  │   → assert_true: Got response expected true got false
  │ FAIL [expected PASS] remote (http://web-platform.test:8000) to remote (http://web-platform.test:8000), expect origin=http://web-platform.test:8000
  │   → assert_true: Got response expected true got false
  │ 
  │ shouldPass/</client.onreadystatechange<@http://web-platform.test:8000/cors/redirect-origin.htm:138:17
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  │ FAIL [expected PASS] remote (*) to remote (*), expect origin=http://web-platform.test:8000
  │   → assert_equals: Origin Header expected (string) "http://web-platform.test:8000" but got (undefined) undefined
  │ FAIL [expected PASS] remote (http://web-platform.test:8000) to remote (*), expect origin=http://web-platform.test:8000
  │   → assert_equals: Origin Header expected (string) "http://web-platform.test:8000" but got (undefined) undefined
  │ 
  │ shouldPass/</client.onreadystatechange<@http://web-platform.test:8000/cors/redirect-origin.htm:140:17
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /cors/redirect-preflight.htm:
  │ FAIL [expected PASS] Should allow redirect 301 after succesful (200) preflight
  │   → assert_equals: Successfull redirect expected 200 but got 0
  │ FAIL [expected PASS] Should allow redirect 302 after succesful (200) preflight
  │   → assert_equals: Successfull redirect expected 200 but got 0
  │ FAIL [expected PASS] Should allow redirect 303 after succesful (200) preflight
  │   → assert_equals: Successfull redirect expected 200 but got 0
  │ FAIL [expected PASS] Should allow redirect 307 after succesful (200) preflight
  │   → assert_equals: Successfull redirect expected 200 but got 0
  │ FAIL [expected PASS] Should allow redirect 308 after succesful (200) preflight
  │   → assert_equals: Successfull redirect expected 200 but got 0
  │ 
  │ redir_after_successfull_preflight/</client.onreadystatechange<@http://web-platform.test:8000/cors/redirect-preflight.htm:39:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /fetch/api/cors/cors-filtering.html:
  └ PASS [expected FAIL] CORS filter on Set-Cookie header

  ▶ Unexpected subtest result in /fetch/api/cors/cors-filtering.html:
  │ FAIL [expected PASS] CORS filter on Set-Cookie2 header
  └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"

  ▶ Unexpected subtest result in /fetch/api/cors/cors-filtering.html:
  └ PASS [expected FAIL] CORS filter on Set-Cookie header, header is forbidden

  ▶ Unexpected subtest result in /fetch/api/cors/cors-filtering.html:
  │ FAIL [expected PASS] CORS filter on Set-Cookie2 header, header is forbidden
  └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"

  ▶ Unexpected subtest result in /fetch/api/cors/cors-filtering-worker.html:
  └ PASS [expected FAIL] CORS filter on Set-Cookie header

  ▶ Unexpected subtest result in /fetch/api/cors/cors-filtering-worker.html:
  │ FAIL [expected PASS] CORS filter on Age header, header is exposed
  └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.html:
  └ PASS [expected FAIL] CORS [DELETE], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.html:
  └ PASS [expected FAIL] CORS [PUT], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.html:
  └ PASS [expected FAIL] CORS [PUT] [only safe headers], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.worker.html:
  └ PASS [expected FAIL] CORS [DELETE], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.worker.html:
  └ PASS [expected FAIL] CORS [PUT], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.worker.html:
  └ PASS [expected FAIL] CORS [PUT] [only safe headers], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-redirect.any.html:
  │ FAIL [expected PASS] Redirect 301: cors to same cors
  │   → assert_equals: Origin is correctly set after redirect expected "http://web-platform.test:8000" but got ""
  │ FAIL [expected PASS] Redirect 302: cors to same cors
  │   → assert_equals: Origin is correctly set after redirect expected "http://web-platform.test:8000" but got ""
  │ FAIL [expected PASS] Redirect 303: cors to same cors
  │   → assert_equals: Origin is correctly set after redirect expected "http://web-platform.test:8000" but got ""
  │ FAIL [expected PASS] Redirect 307: cors to same cors
  │   → assert_equals: Origin is correctly set after redirect expected "http://web-platform.test:8000" but got ""
  │ FAIL [expected PASS] Redirect 308: cors to same cors
  │   → assert_equals: Origin is correctly set after redirect expected "http://web-platform.test:8000" but got ""
  │ 
  └ corsRedirect/</</<@http://web-platform.test:8000/fetch/api/cors/cors-redirect.any.js:19:9

  ▶ TIMEOUT [expected OK] /eventsource/eventsource-cross-origin.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ Unexpected subtest result in /eventsource/eventsource-cross-origin.htm:
  │ TIMEOUT [expected PASS] EventSource: cross-origin redirect use
  └   → Test timed out
```

**Comment by KiChjang** [comment]
@jdm I'm not sure what to make of these test results. Comparing the request's current URL's origin with the location URL's origin is actually *more* correct than using the request's origin for comparison.

**Comment by jdm** [comment]
More correct in terms of tests that pass? That just suggests to me that our request's origin is not what we expect, which means we should fix it :)

**Comment by KiChjang** [comment]
```
  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.html:
  └ PASS [expected FAIL] CORS [DELETE], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.html:
  └ PASS [expected FAIL] CORS [PUT], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.html:
  └ PASS [expected FAIL] CORS [PUT] [only safe headers], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.worker.html:
  └ PASS [expected FAIL] CORS [DELETE], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.worker.html:
  └ PASS [expected FAIL] CORS [PUT], server allows

  ▶ Unexpected subtest result in /fetch/api/cors/cors-preflight.any.worker.html:
  └ PASS [expected FAIL] CORS [PUT] [only safe headers], server allows
```

**Comment by KiChjang** [comment]
Mind if I file the investigation as a follow-up? I'd rather land this first.

**Comment by jdm** [comment]
Sure. I have a suspicion that we end up using Origin::Client instead of a real origin.

**Comment by jdm** [comment]
This can r=me with the WIP commit merged into the original related change.

**Comment by KiChjang** [comment]
@bors-servo retry
* infra

**Pull request merged**